bitcash/network/services.py

```diff
@@ -20,215 +20,43 @@ def set_service_timeout(seconds):
-        return response
-
-    @classmethod
-    def broadcast_tx(cls, tx_hex, network):  # pragma: no cover
-        API = cls.network_endpoint(network) + cls.RAW_TX_PATH
-        r = requests.get(API.format(tx_hex))
-        return True if r.status_code == 200 else False
-
-
-class BitcoreAPI:
-    """ Insight API v8 """
-
-    NETWORK_ENDPOINTS = {
-        "mainnet": os.getenv(
-            "BITCORE_API_MAINNET", "https://api.bitcore.io/api/BCH/mainnet/"
-        ),
-        "testnet": os.getenv(
-            "BITCORE_API_TESTNET", "https://api.bitcore.io/api/BCH/testnet/"
-        ),
-    }
-
-    MAIN_ENDPOINT = "https://api.bitcore.io/api/BCH/mainnet/"
-    ADDRESS_API = "address/{}"
-    BALANCE_API = ADDRESS_API + "/balance"
-    UNSPENT_API = ADDRESS_API + "/?unspent=true"
-    TX_PUSH_API = "tx/send"
-    TX_API = "tx/{}"
-    TX_AMOUNT_API = TX_API
-    TX_PUSH_PARAM = "rawTx"
-
-    @classmethod
-    def network_endpoint(cls, network):
-        if network not in NETWORKS:
-            raise InvalidNetwork(f"No endpoints found for network {network}")
-        return cls.NETWORK_ENDPOINTS[network]
-
-    @classmethod
-    def remove_prefix(cls, address):
-        if ":" in address:
-            address = address.split(":")[1]
-        return address
-        # else:
-        #     raise InvalidAddress(address)
-
-    @classmethod
-    def get_unspent(cls, address, network):
-        address = cls.remove_prefix(address)
-        API = cls.network_endpoint(network) + cls.UNSPENT_API
-        r = requests.get(API.format(address), timeout=DEFAULT_TIMEOUT)
-        r.raise_for_status()
-        unspents = []
-        for tx in r.json():
-            # In weird conditions, the API will send back unspents
-            # without a scriptPubKey.
-            if "script" in tx:
-                unspents.append(
-                    Unspent(
-                        currency_to_satoshi(tx["value"], "satoshi"),
-                        tx["confirmations"],
-                        tx["script"],
-                        tx["mintTxid"],
-                        tx["mintIndex"],
+def get_endpoints_for(network):
+    # For each available interface in 'ENDPOINT_ENV_VARIABLES'
+    # this function will check, in order, if any env variables
+    # have been set for EITHER:
+    # <NAME>_API_<NETWORK>
+    # OR
+    # <NAME>_API_<NETWORK>_<N>
+    # Where 'N' is a number starting at 1 and increasing to
+    # however many endpoints you'd like.
+    # If neither of these env variables have been set, it returns
+    # the instantiated result of <NAME>.get_default_endpoints(network)
+
+    endpoints = []
+    for endpoint in ENDPOINT_ENV_VARIABLES.keys():
+        if os.getenv(f"{endpoint}_API_{network}".upper()):
+            endpoints.append(
+                ENDPOINT_ENV_VARIABLES[endpoint](
+                    os.getenv(f"{endpoint}_API_{network}".upper())))
+        elif os.getenv(f"{endpoint}_API_{network}_1".upper()):
```

Comment:
This is definitely clever and I'm okay with it. However, I tend to think that environment variables to configure Python isn't the best practice.

I think I'd rather have some type of override for the endpoints, maybe with the class. I like environment variables to be upstream. So the application can take environment variables, but the libraries, I don't think so.

That said, this definitely works and I'm fine with it. I think you did a nice job with it.

Reply:
Personally I work with Docker containers quite often, so environmental variables are one of my "go-to"s for configuration options. I'm fine if this gets changed to some other mechanism, but the design @merc1er and I worked out was to do it this way.